### ISSUE TITLE
Some hex values mistakenly parsed as zeroes

### ISSUE THREAD

**Issue by vladikcomper** [r1chardj0n3s/parse#83]
For some reason, parsing certain hex values with leading zeroes produces buggy and unreliable results.
Here's a quick demonstration using the latest version of parse (1.12.0):

```
>>> from parse import parse
>>> parse.parse('${:x}','$0b67')
<Result (0,) {}>
>>> parse.parse('${:x}','$0B67')
<Result (0,) {}>
>>> parse.parse('${:x}','$0B6')
<Result (0,) {}>
>>> parse.parse('${:x}','$0B')
<Result (11,) {}>
>>> parse.parse('${:x}','$B67')
<Result (2919,) {}>
```

It appears that parse() recognizes, but fails to parse certain combinations of hex digits, returning zero instead.
Interestingly, whether parsing fails depends on the digit next to zero.
Based on my testing, it always happens with numbers starting with "0Bxxx" (excluding "0B"; I know there was a separate [closed] bug report on that one, but it appears that the underlying issue is still there).

**Comment by vladikcomper** [comment]
Well, I see the roots of it now.
It recognizes "0b" as binary number prefix, disregarding the actual type. 
Encoding pure binary values inside the hex numbers further reveals the actual bug, resulting in completely wrong parsing:

```
>>> parse.parse('${:x}','$0b11')
<Result (3,) {}>
>>> parse.parse('${:x}','$0b110')
<Result (6,) {}>
```
Obviously, the results should be 2833 and 45328 correspondingly in the example above.

**Comment by vladikcomper** [comment]
The bug is with **_int_convert()_** function.

Altering **_int_convert()_** in the following way fixes the issue for me:

```
def int_convert(base):
    '''Convert a string to an integer.

    The string may start with a sign.

    It may be of a base other than 10.

    If may start with a base indicator, 0#nnnn, which we assume should
    override the specified base.

    It may also have other non-numeric characters that we can ignore.
    '''
    CHARS = '0123456789abcdefghijklmnopqrstuvwxyz'

    def f(string, match, base=base):
        if string[0] == '-':
            sign = -1
        else:
            sign = 1

        # Try to guess the base if the base wasn't specified
        if base is None:
            if string[0] == '0' and len(string) > 2:
                if string[1] in 'bB':
                    base = 2
                elif string[1] in 'oO':
                    base = 8
                elif string[1] in 'xX':
                    base = 16
                else:
                    # Otherwise, just assume decimal number ...
                    base = 10 

        chars = CHARS[:base]
        string = re.sub('[^%s]' % chars, '', string.lower())
        return sign * int(string, base)
    return f
```

I'm not even sure base detection logic is necessary, because every time it gets called, the base _is already_ specified. 
So the function only breaks correctly passed base on certain occasions, like if the number to convert starts with '0b'.

**Comment by r1chardj0n3s** [comment]
Thanks for the detective work!

**Comment by cantzler** [comment]
Oh man this just bit me as well -- nefarious bug causing any hex "0b" value to be read as zero.
The proposed fix for adding "if base is None:" worked for me.

Can this make a production release of the package please?

**Comment by cantzler** [comment]
Here's workaround folks can use (since I need a deploy-able w/a)

`
    test = parse.parse('{:x}','0b0101')
    if test.fixed[0] != 0xb0101:
        # need to workaround bug were parse guesses base (when it shouldn't)
        # https://github.com/r1chardj0n3s/parse/issues/83
        import re
        def int_convert(base):
            CHARS = '0123456789abcdefghijklmnopqrstuvwxyz'
            def f(string, match, base=base):
                if string[0] == '-':
                    sign = -1
                else:
                    sign = 1
                # Try to guess the base if the base wasn't specified
                if base is None:
                    if string[0] == '0' and len(string) > 2:
                        if string[1] in 'bB':
                            base = 2
                        elif string[1] in 'oO':
                            base = 8
                        elif string[1] in 'xX':
                            base = 16
                        else:
                            # Otherwise, just assume decimal number ...
                            base = 10 
                chars = CHARS[:base]
                string = re.sub('[^%s]' % chars, '', string.lower())
                return sign * int(string, base)
            return f
        parse.int_convert = int_convert

    test = parse.parse('{:x}','0b0101')
    if test.fixed[0] != 0xb0101:
        raise Exception("ERROR! Unable to patch 'parse' for known issue: https://github.com/r1chardj0n3s/parse/issues/83")
`

**Comment by vladikcomper** [comment]
@cantzler, I've just created a PR #103 that fixes this issue and a few other bugs I discovered when implementing my original fix (also #102 for more information on these bugs).
You may use the proposed patches in production, as they were properly tested and make parsing of non-decimal numbers less buggy.
